addon/duckdb.cc

```diff
@@ -61,62 +60,157 @@ DuckDB::DuckDB(const Napi::CallbackInfo &info)
+    open_file_callback_ref = Napi::Persistent(
+        file_system_object.Value().Get("openFile").As<Napi::Function>());
+    truncate_callback_ref = Napi::Persistent(
+        file_system_object.Value().Get("truncate").As<Napi::Function>());
+
+    read_with_location_callback_tsfn =
+        Napi::ThreadSafeFunction::New(read_with_location_callback_ref.Env(),
+                                      read_with_location_callback_ref.Value(),
+                                      "read_with_location_callback_tsfn",
+                                      0, // Unlimited queue
+                                      1);
+    read_tsfn = Napi::ThreadSafeFunction::New(
+        read_callback_ref.Env(), read_callback_ref.Value(), "read_tsfn",
+        0, // Unlimited queue
+        1);
+    glob_tsfn = Napi::ThreadSafeFunction::New(
+        glob_callback_ref.Env(), glob_callback_ref.Value(), "glob_tsfn",
+        0, // Unlimited queue
+        1);
+    get_file_size_tsfn = Napi::ThreadSafeFunction::New(
+        get_file_size_callback_ref.Env(), get_file_size_callback_ref.Value(),
+        "get_file_size_tsfn",
+        0, // Unlimited queue
+        1);
+    open_file_tsfn = Napi::ThreadSafeFunction::New(
+        open_file_callback_ref.Env(), open_file_callback_ref.Value(),
+        "open_file_tsfn",
+        0, // Unlimited queue
+        1);
+    truncate_tsfn = Napi::ThreadSafeFunction::New(truncate_callback_ref.Env(),
+                                                  truncate_callback_ref.Value(),
+                                                  "truncate_tsfn",
+                                                  0, // Unlimited queue
+                                                  1);
+
+    nativeConfig.file_system = duckdb::make_unique<NodeFileSystem>(
+        read_with_location_callback_tsfn, read_tsfn, glob_tsfn,
+        get_file_size_tsfn, open_file_tsfn, truncate_tsfn);
+  }
+}
+
+Napi::Value DuckDB::Init(const Napi::CallbackInfo &info) {
+  if (initialized) {
+    throw Napi::Error::New(info.Env(), "Has already been initialized");
+  }
+  if (!info[0].IsFunction()) {
+    throw Napi::TypeError::New(info.Env(),
+                               "Invalid argument: must be an function");
+  }
+  initialized = true;
+  init_tsfn = Napi::ThreadSafeFunction::New(
+      info.Env(),
+      info[0].As<Napi::Function>(), // JavaScript function called asynchronously
+      "DuckDB Init",                // Name
+      0,                            // Unlimited queue
+      1,                            // Only one thread will use this initially
+      [&](Napi::Env) {              // Finalizer used to clean threads up
+        nativeThread.join();
+      });
+
+  nativeThread = std::thread([&] {
+    database = duckdb::make_unique<duckdb::DuckDB>(path, &nativeConfig);
+    database->LoadExtension<duckdb::ParquetExtension>();
+    init_tsfn.BlockingCall();
+    init_tsfn.Release();
+  });
+  return info.Env().Undefined();
```

Comment:
Could be nice to return the DuckDB instance here for single line instantiation, e.g. 

```ts
const duckdb = await new DuckDB().init();
```